connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java

```diff
@@ -279,10 +284,61 @@ public Future<Void> set(Map<ByteBuffer, ByteBuffer> values, Callback<Void> callb
             throw new IllegalStateException("At least one non-null offset store must be provided");
         }
 
+        boolean containsTombstones = values.containsValue(null);
+
+        // If there are tombstone offsets, then the failure to write to secondary store will
+        // not be ignored. Also, for tombstone records, we first write to secondary store and
+        // then to primary stores.
+        if (secondaryStore != null && containsTombstones) {
+            AtomicReference<Throwable> secondaryStoreTombstoneWriteError = new AtomicReference<>();
+            Future<Void> secondaryWriteFuture = secondaryStore.set(values, (secondaryWriteError, ignored) -> {
+                try (LoggingContext context = loggingContext()) {
+                    if (secondaryWriteError != null) {
+                        log.warn("Failed to write offsets with tombstone records to secondary backing store", secondaryWriteError);
+                        secondaryStoreTombstoneWriteError.compareAndSet(null, secondaryWriteError);
+                    } else {
+                        log.debug("Successfully flushed tombstone offsets to secondary backing store");
+                    }
+                }
+            });
```

Comment:
Why do we need to do this if we're anyway doing a get on the returned future immediately? Any producer error will be wrapped in an `ExecutionException` and thrown when we call get on the future returned from `KafkaOffsetBackingStore` (the same exception will also be used to complete the passed callback exceptionally) - https://github.com/apache/kafka/blob/546b912b831d46208499c6e1f4f785db6ddd9aa3/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java#L344-L414

Reply:
I see what you are saying. I think what you mean is that the callback passed to the `set` method itself can be provided to the `KafkaOffsetBackingStore#set` method and it will take care of completing it when the `onComplete` of `SetCallback` is completed. I think that makes sense but the reason I did this was that I can't return the `Future<Void>` object created when invoking `KafkaOffsetBackingStore#set` object always because if that operation succeeds then we also need to write to the primary store. That's why, created this extra callback which gets executed upon `onComplete` of `SetCallback` and with this I can selectively return a `FutureCallBack<Void>` object by giving it the original callback. Let me know if that makes sense?

Reply:
I tried something like this =>
```
FutureCallback<Void> failedWriteCallback = new FutureCallback<>(callback);
            secondaryStore.set(values, failedWriteCallback);
/*
            Future<Void> secondaryWriteFuture = secondaryStore.set(values, (secondaryWriteError, ignored) -> {
                try (LoggingContext context = loggingContext()) {
                    if (secondaryWriteError != null) {
                        log.warn("Failed to write offsets with tombstone records to secondary backing store", secondaryWriteError);
                        secondaryStoreTombstoneWriteError.compareAndSet(null, secondaryWriteError);
                    } else {
                        log.debug("Successfully flushed tombstone(s)-containing to secondary backing store");
                    }
                }
            });
*/
            try {
                if (exactlyOnce) {
                    failedWriteCallback.get();
                } else {
                    failedWriteCallback.get(offsetFlushTimeoutMs, TimeUnit.MILLISECONDS);
                }
```
which does whatever we discussed above i.e the SetCallback, when it's `onComplete` is executed, executes the `onComplete` of the `FutureCallback` object which eventually executes the actual passed in `callback`. But, when the secondary write fails, we need to eventually return another `Future` object . Maybe I am still missing the point you are making here. We can sync up maybe on this.

Reply:
I made the change. That has worked. Thanks!

Reply:
I was suggesting the use of a plain `FutureCallback` instantiated via the default constructor. We're still creating an unnecessary callback to handle `secondaryWriteError` which will anyway be surfaced synchronously when we call `get` on our `FutureCallback`?